src/btree/bt_discard.c

```diff
@@ -245,20 +245,31 @@ __free_page_modify(WT_SESSION_IMPL *session, WT_PAGE *page)
 void
 __wt_ref_addr_free(WT_SESSION_IMPL *session, WT_REF *ref)
 {
+    WT_PAGE *home;
     void *ref_addr;
 
     /*
-     * The page being discarded may be the child of a page being split, where the WT_REF.addr field
-     * is being instantiated (as it can no longer reference the on-disk image). Loop until we read
-     * and clear the address without a race, then free the read address as necessary.
+     * In order to free the WT_REF.addr field we need to read and clear the address without a race.
+     * The WT_REF may be a child of a page being split, in which case the addr field could be
+     * instantiated concurrently which changes the addr field. Once we swap in NULL we effectively
+     * own the addr. Then provided the addr is off page we can free the memory.
+     *
+     * However as we could be the child of a page being split the ref->home pointer which tells us
+     * whether the addr is on or off page could change concurrently. To avoid this we save the home
+     * pointer before we do the compare and swap. Additionally ref->home is marked as volatile so
+     * the compiler won't be able to replace this with multiple reads throughout the function.
      */
+    home = ref->home;
     do {
```

Comment:
Does `home` need to be re-read within this loop? Is it technically possible for the home to be updated multiple times while this thread is trying to free a ref?

Reply:
My intuition is that no it doesn't, if home is updated then the address will have to have been instantiated by the thread that performed the `split_ref_move` and the `ref->addr` field is then pointing at an instantiated structure. As such we don't mind if we have the wrong home as it should be guaranteed to be "off page".

Reply:
I guess the only possible other outcome is that the older home got freed, and then the instantiated address actually got allocated in the same segment of memory and thus we incorrectly assume it to be off page?

Reply:
Upon moving to a new parent the `home->dsk` will be NULL which would prevent us from seeing it as "on page".

Reply:
[52817b9](https://github.com/wiredtiger/wiredtiger/pull/8870/commits/52817b9aeb4a08e15433f58bf2f17c5012626eb8)